src/libcore/iter/range.rs

```diff
@@ -421,6 +421,19 @@ impl<A: Step> Iterator for ops::RangeInclusive<A> {
         Try::from_ok(accum)
     }
 
+    #[inline]
+    fn fold<B, F>(mut self, init: B, f: F) -> B
+    where
+        Self: Sized,
+        F: FnMut(B, Self::Item) -> B,
+    {
+        #[inline]
+        fn ok<B, T>(mut f: impl FnMut(B, T) -> B) -> impl FnMut(B, T) -> Result<B, !> {
+            move |acc, x| Ok(f(acc, x))
+        }
+
+        self.try_fold(init, ok(f)).unwrap()
```

Comment:
I tried this:
```
    #[inline]
    fn fold<B, F>(mut self, init: B, mut f: F) -> B
    where
        Self: Sized,
        F: FnMut(B, Self::Item) -> B,
    {
        let mut accum = init;

        while self.start <= self.end {
            let n = Step::forward(self.start.clone(), 1);
            let n = mem::replace(&mut self.start, n);
            accum = f(accum, n);
        }

        // No need to set `self.exhausted` because `self` is consumed.

        accum
    }
```
But it made `src/test/codegen/issue-45222.rs` fail :(  Hard to believe that is less optimizable than `try_fold`, but what can you do. So I've left that part unchanged.